omnivector-solutions/slurmdbd-operator
# Configure database per the DatabaseCreatedEvent

## Description

The DatabaseCreatedEvent handler took a rather simple approach to defining database connections and always used a local TCP connection to the MySQL Router on a standard port. This worked at the time, but depended on behavior which has since changed. Instead, this patch changes the _on_database_created event handler to process the endpoints and configure the slurmdbd service based upon the the endpoints.

The database relation is a generic relation and allows for the slurmdbd charm to be connected to either a mysql-router charm or to a mysql service, depending on whether the database service is configured for high availability or not. The changes in this patch handle both scenarios, ensuring that the charm properly attempts to configure itself as directed by the relation data.

In a highly available configuration, the mysql-router subordinate charm will be configured for connections and the slurmdbd service will use the unix socket for communication. This involves specifying environment variables to the slurmdbd process to indicate which socket to use.

If the slurmdbd charm is related directly to mysql then the slurmdbd service will connect over a TCP socket. In this case, the slurmdbd.conf file can be updated to point to the host address and port that the mysql service is listening on.

## How was the code tested?

Code was tested in multiple ways:
 - [X] Unit Tests (new and existing)
 - [X] Integration Tests (existing)
   - Log into machines, verify all the necessary bits are properly set
   - Verify that services are running
   - Verify that database schemas are created

Operating Systems Tested on:
 - [X] Ubuntu 22.04
 - [ ] Ubuntu 20.04
 - [ ] Centos 7

## Related issues and/or tasks

This PR depends on https://github.com/omnivector-solutions/slurm-ops-manager/pull/97. It will not run integration tests successfully until the slurmdbd configuration template has the option to skip defining the StorageHost and StoragePort options.

*Note: The referenced change in the slurm-ops-manager may need to be backported as the slurmdbd-operator currently depends on the 0.8.12 branch of slurm-ops-manager.*

## Checklist

- [X] I am the author of these changes, or I have the rights to submit them.
- [X] I have added the relevant changes to the README and/or documentation.
- [ ] All requested changes and/or review comments have been resolved.
